File: 02_Object_Oriented_Programming/01_Introduction_to_Classes_And_Objects/javascript/car.js
/**
 * car.js - Demonstrates Classes and Objects in JavaScript (ES6+)
 *
 * Key Concepts:
 *   - ES6 class syntax with constructor
 *   - Private-like fields using # (true private fields — ES2022)
 *   - Getters and setters
 *   - Methods that operate on internal state
 *   - toString() override
 */

class Car {
  // ── Private fields (ES2022) ─────────────────────────────────────────
  #manufacturer;
  #model;
  #year;
  #engineRunning;
  #speed;

  // ── Constructor ─────────────────────────────────────────────────────
  constructor(manufacturer, model, year) {
    this.#manufacturer = manufacturer;
    this.#model = model;
    this.#year = year;
    this.#engineRunning = false;
    this.#speed = 0;
  }

  // ── Getters ─────────────────────────────────────────────────────────
  get manufacturer() {
    return this.#manufacturer;
  }

  get model() {
    return this.#model;
  }

  get year() {
    return this.#year;
  }

  get engineRunning() {
    return this.#engineRunning;
  }

  get speed() {
    return this.#speed;
  }

  // ── Setters (only where mutation makes sense) ───────────────────────
  set manufacturer(value) {
    this.#manufacturer = value;
  }

  set model(value) {
    this.#model = value;
  }

  set year(value) {
    this.#year = value;
  }

  // ── Methods ─────────────────────────────────────────────────────────
  startEngine() {
    if (this.#engineRunning) {
      console.log(`${this} → Engine is already running.`);
      return;
    }
    this.#engineRunning = true;
    console.log(`${this} → Engine started.`);
  }

  stopEngine() {
    if (!this.#engineRunning) {
      console.log(`${this} → Engine is already off.`);
      return;
    }
    if (this.#speed > 0) {
      console.log(
        `${this} → Cannot stop engine while moving! Current speed: ${this.#speed} km/h`
      );
      return;
    }
    this.#engineRunning = false;
    console.log(`${this} → Engine stopped.`);
  }

  accelerate(increment) {
    if (!this.#engineRunning) {
      console.log(`${this} → Start the engine first!`);
      return;
    }
    this.#speed += increment;
    console.log(`${this} → Accelerated to ${this.#speed} km/h.`);
  }

  brake(decrement) {
    if (this.#speed === 0) {
      console.log(`${this} → Car is already stationary.`);
      return;
    }
    this.#speed = Math.max(0, this.#speed - decrement);
    console.log(`${this} → Braked to ${this.#speed} km/h.`);
  }

  displayInfo() {
    const engineStatus = this.#engineRunning ? "ON" : "OFF";
    console.log(
      `Car Info: ${this.#manufacturer} ${this.#model} (${this.#year})` +
        ` | Engine: ${engineStatus}` +
        ` | Speed: ${this.#speed} km/h`
    );
  }

  // ── toString ────────────────────────────────────────────────────────
  toString() {
    return `${this.#year} ${this.#manufacturer} ${this.#model}`;
  }
}

module.exports = Car;

File: 02_Object_Oriented_Programming/01_Introduction_to_Classes_And_Objects/java/Car.java
/**
 * Car.java - Demonstrates Classes and Objects in Java
 *
 * Key Concepts:
 *   - Class definition with attributes (fields) and methods
 *   - Constructor for initialising object state
 *   - Encapsulation via private fields + public getters/setters
 *   - Method behaviour tied to object state
 *   - Overriding toString() for readable output
 */
public class Car {

    // ── Attributes (private for encapsulation) ──────────────────────────
    private String manufacturer;
    private String model;
    private int year;
    private boolean engineRunning;
    private int speed; // km/h

    // ── Constructor ─────────────────────────────────────────────────────
    public Car(String manufacturer, String model, int year) {
        this.manufacturer = manufacturer;
        this.model = model;
        this.year = year;
        this.engineRunning = false;
        this.speed = 0;
    }

    // ── Getters ─────────────────────────────────────────────────────────
    public String getManufacturer() {
        return manufacturer;
    }

    public String getModel() {
        return model;
    }

    public int getYear() {
        return year;
    }

    public boolean isEngineRunning() {
        return engineRunning;
    }

    public int getSpeed() {
        return speed;
    }

    // ── Setters (only where mutation makes sense) ───────────────────────
    public void setManufacturer(String manufacturer) {
        this.manufacturer = manufacturer;
    }

    public void setModel(String model) {
        this.model = model;
    }

    public void setYear(int year) {
        this.year = year;
    }

    // ── Behaviours / Methods ────────────────────────────────────────────
    public void startEngine() {
        if (engineRunning) {
            System.out.println(this + " → Engine is already running.");
            return;
        }
        engineRunning = true;
        System.out.println(this + " → Engine started.");
    }

    public void stopEngine() {
        if (!engineRunning) {
            System.out.println(this + " → Engine is already off.");
            return;
        }
        if (speed > 0) {
            System.out.println(this + " → Cannot stop engine while moving! Current speed: " + speed + " km/h");
            return;
        }
        engineRunning = false;
        System.out.println(this + " → Engine stopped.");
    }

    public void accelerate(int increment) {
        if (!engineRunning) {
            System.out.println(this + " → Start the engine first!");
            return;
        }
        speed += increment;
        System.out.println(this + " → Accelerated to " + speed + " km/h.");
    }

    public void brake(int decrement) {
        if (speed == 0) {
            System.out.println(this + " → Car is already stationary.");
            return;
        }
        speed = Math.max(0, speed - decrement);
        System.out.println(this + " → Braked to " + speed + " km/h.");
    }

    public void displayInfo() {
        System.out.println("Car Info: " + manufacturer + " " + model + " (" + year + ")"
                + " | Engine: " + (engineRunning ? "ON" : "OFF")
                + " | Speed: " + speed + " km/h");
    }

    // ── toString ────────────────────────────────────────────────────────
    @Override
    public String toString() {
        return year + " " + manufacturer + " " + model;
    }
}
